### PR TITLE
Fix issue with cache references improperly combined with control flow

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
1. Fixes control flow caching issue.
The array cache manager returns references to arrays that need to be reused in subsequent iterations of  a loop. With the cache enabled we may use arrays in outputs which can break subsequent iterations of a loop variable.

2. Adds multinomial op for import.

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [ X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X ] Created tests for any significant new code additions.
- [ X] Relevant tests for your changes are passing.

### CODE REVIEW

**Review by quickwritereader — DISMISSED** [review]

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/AbstractSession.java [hunk 1]
```diff
@@ -505,23 +504,18 @@ else if(otherPlaceHolderValues.containsKey(es.getName())) {
...
-                                    if(tensorArraysInSession != null) {
-                                        tensorArraysInSession.addAll(sdValue.getListValue());
-                                    } else {
-                                        tensorArrays.put(vid, (List<T>) sdValue.getListValue());
-                                    }
-                                    //note: we omit break on purpose
+                                    //note: we leave this out since we already update node value outputs earlier
+                                    putNodeValue(sdValue,vid);
+                                    break;
+                                //note: we omit break on purpose
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/AbstractSession.java]: This doesn't make sense. Do we omit break on purpose, or do we break?

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/InferenceSession.java [hunk 2]
```diff
@@ -404,16 +417,22 @@ else if(out.hasValues()) {
...
+                            }
+                            break;
+                        case LIST:
+                            for(INDArray arr : value.getListValue())
+                                if(arr != null && !freedArrays.contains(arr.getId())) {
+                                    mmgr.release(arr);
+                                    freedArrays.add(arr.getId());
+                                }
+                            break;
+                    }*/
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/InferenceSession.java]: Please don't leave any commented out code behind.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/InferenceSession.java [hunk 3]
```diff
@@ -849,22 +909,19 @@ public ExecutionResult getOutputsHelperTensorArrayOps(DifferentialFunction op, F
...
 
+
+
             //create new tensor array for placeholder referencing a passed in variable
             if(tArr == null && inTensorArray.getVariableType() == VariableType.PLACEHOLDER) {
                 VarId varId = new VarId(inTensorArray.name(),outputFrameIter.getFrame(),outputFrameIter.getIteration(),outputFrameIter.getParentFrame());
                 tArr = varId;
                 SDValue sdValue = otherPlaceHolders.get(inTensorArray.name());
-                tensorArrays.put(tArr,sdValue.getListValue());
+                //putNodeValue(sdValue, tArr);
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/InferenceSession.java:919]: commented out code

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/InferenceSession.java [hunk 4]
```diff
@@ -880,27 +937,39 @@ public ExecutionResult getOutputsHelperTensorArrayOps(DifferentialFunction op, F
...
+            log.trace("Setting item at index " + idx + " for list " + tArr + " with value " + arr + " with whole list of after write " + l + " and value array " + arr);
+            log.trace("Writing value " + inSDV + " to list " + tArr.getVariable() + " at iteration " + tArr.getIteration());
 
             //Add a dependency
             Dep d = new ExecDoneDep();
-            VarId varId = new VarId(op.outputVariable().name(), outputFrameIter.getFrame(),outputFrameIter.getIteration(),outputFrameIter.getParentFrame());
-            nodeValueOutputs.put(varId,nodeValueOutputs.get(tArr));
-            arrayUseTracker.addDependency(nodeValueOutputs.get(tArr), d);
-            return ExecutionResult.createValue(op.outputVariable().name(),nodeValueOutputs.get(tArr));
+            //putNodeValue(sdValue1, tArr);
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/InferenceSession.java]: commented out code

File: platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/autodiff/samediff/MemoryMgrTest.java [hunk 5]
```diff
@@ -51,7 +51,7 @@ public char ordering(){
     @MethodSource("org.nd4j.linalg.BaseNd4jTestWithBackends#configs")
     public void testArrayReuseTooLarge(Nd4jBackend backend) throws Exception {
 
-        ArrayCacheMemoryMgr mmgr = new ArrayCacheMemoryMgr();
+    /*    ArrayCacheMemoryMgr mmgr = new ArrayCacheMemoryMgr();
```
Comment (treo) [inline platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/autodiff/samediff/MemoryMgrTest.java]: commented out code

**Review by treo — CHANGES_REQUESTED** [review]

**Review by treo — APPROVED** [review]

**Pull request merged**